### PR TITLE
HTTP Basic Auth implementation

### PR DESCRIPTION
**Pull request by saiHemak**
This PR is based on  #1407 . I have picked up the implementation from #1407  PR and added HTTPServer changes which are required for testing this functionality.

@pushkarnk please review

### CODE REVIEW

File: Foundation/URLAuthenticationChallenge.swift [hunk 1]
```diff
@@ -192,3 +192,21 @@ open class URLAuthenticationChallenge : NSObject, NSSecureCoding {
         }
     }
 }
+
+extension _HTTPURLProtocol : URLAuthenticationChallengeSender {
```
Comment (pushkarnk) [inline Foundation/URLAuthenticationChallenge.swift]: We must add `NSUnimplemented()` or `fatalError()` with a message to these empty methods.

**Comment by pushkarnk** [comment]
@swift-ci please test

File: Foundation/URLAuthenticationChallenge.swift (hunk 1 above)
Comment (saiHemak) [reply Foundation/URLAuthenticationChallenge.swift]: Done ..Thanks

**Comment by saiHemak** [comment]
@pushkarnk  As of now the code is crashing when the header ```"WWW-Authenticate``` is not present.So, we can handle this in two ways either throwing the ```fatalError``` with a message which will be no different from the earlier crash but with a proper message or invoking the ```urlProtocol(`protocol`, didFailWithError:``` delegate . I have implemented the latter way . However the error message relevant to this scenario seems to be unavailable in ```URLError``` class. I am not sure whether I can add new errors to this class . Hence I have set the  error code as ```NSURLErrorUserAuthenticationRequired``` . Please let me know  if this is acceptable.

File: Foundation/URLSession/URLSessionTask.swift [hunk 2]
```diff
@@ -580,7 +616,24 @@ extension _ProtocolClient : URLProtocolClient {
     }
 
     func urlProtocol(_ protocol: URLProtocol, didReceive challenge: URLAuthenticationChallenge) {
-        NSUnimplemented()
+        guard let task = `protocol`.task else { fatalError() }
```
Comment (pushkarnk) [inline Foundation/URLSession/URLSessionTask.swift]: Please call `fatalError()` with a  clear error message.

File: Foundation/URLSession/URLSessionTask.swift [hunk 3]
```diff
@@ -580,7 +616,24 @@ extension _ProtocolClient : URLProtocolClient {
...
+        guard let session = task.session as? URLSession else { fatalError() }
```
Comment (pushkarnk) [inline Foundation/URLSession/URLSessionTask.swift]: Please call fatalError() with a clear error message.

**Comment by pushkarnk** [comment]
@saiHemak Please correct the typo in the commit message as well. Thanks.

**Comment by pushkarnk** [comment]
@swift-ci please test

**Comment by pushkarnk** [comment]
@swift-ci please test and merge

**Comment by pushkarnk** [comment]
@swift-ci please test and merge

**Comment by pushkarnk** [comment]
@swift-ci please test and merge

**Comment by pushkarnk** [comment]
@swift-ci please test and merge

**Comment by pushkarnk** [comment]
@swift-ci please test and merge

**Pull request merged**